### PR TITLE
FEATURE: Add last visit indication to topic view page.

### PR DESCRIPTION
**Pull request by tgxworld**
### Desktop

Screenshot from 2021-06-23 15-14-21

### Mobile

Screenshot from 2021-06-28 13-38-14

### CODE REVIEW

File: app/assets/javascripts/discourse/app/widgets/post-stream.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/widgets/post-stream.js:283]: This is to prevent the last visit line from shifting as new posts come in while the user is on the page.

File: app/assets/stylesheets/common/base/topic-post.scss
Comment (jjaffeux) [inline app/assets/stylesheets/common/base/topic-post.scss:1039]: Could we have a class for this please? I expect people might want to customise it.

**Comment by SamSaffron** [comment]
I would like this change to run a bit deeper, I think part of it should be cleaning up `topic_users`, we no longer need both `last_read_post_number` and `highest_seen_post_number` we can drop the `highest_seen_post_number` column together with all the complexity that goes with it.

**Comment by tgxworld** [comment]
@SamSaffron I'll prefer to do the server side clean up in another PR. Does that sound OK to you?

File: app/assets/stylesheets/common/base/topic-post.scss
Comment (tgxworld) [reply app/assets/stylesheets/common/base/topic-post.scss:1039]: People can target it with `.topic-post-visited-line span` if they want to. Also, I'll prefer if we add the class only when there is a need for it. We have `.topic-list-item-separator` with a `td span` element as well for the topic stream.

Comment (jjaffeux) [reply app/assets/stylesheets/common/base/topic-post.scss:1039]: I'm sorry but no, if they want to style it, we need a class, directly styling elements cause lots of issues later in theming. This is an element which might get restyle and a such as "the need for it".

Comment (tgxworld) [reply app/assets/stylesheets/common/base/topic-post.scss]: I've added a class for the `span.`

File: app/assets/javascripts/discourse/app/models/topic.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/models/topic.js:288]: This doesn't make sense to me in the current time and age of reading the code since the session prop `highestSeenByTopic` is set asynchronously while flushing our recently read data.

File: app/assets/javascripts/discourse/app/widgets/post-stream.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/widgets/post-stream.js:282]: Don't show last visit line on last post.

**Comment by jjaffeux** [comment]
The thing am the most worried with this, is people overwriting templates. We should do a check on themes/plugins of big users to see if this will have any impact.

**Comment by tgxworld** [comment]
@jjaffeux Yea it looks like we have a bunch of internal themes that are overriding templates. I'll submit PRs for them.

File: db/post_migrate/20210624023831_remove_highest_seen_post_number_from_topic_users.rb
Comment (martin-brennan) [inline db/post_migrate/20210624023831_remove_highest_seen_post_number_from_topic_users.rb:5]: I am just wondering if it might be best to do this in a separate PR to make sure everything is all good? Otherwise we cannot get the data back if we need to for testing or whatever. Not a huge problem to leave it, just wondering.

File: spec/fixtures/json/sam-s-simple-theme.dcstyle.json
Comment (martin-brennan) [inline spec/fixtures/json/sam-s-simple-theme.dcstyle.json:1]: What has even changed here haha?

**Review by martin-brennan — COMMENTED** [review]
Looks great and from what I can tell it is working fine locally. Just a couple of minor questions. It seems like all the SQL changes have been made accurately and tests are passing so good sign!

You might want to check on the tracking changes in https://github.com/discourse-org/discourse-teams-sidebar as well. From what I can see they should be fine, just another set of eyes would be good.

File: db/post_migrate/20210624023831_remove_highest_seen_post_number_from_topic_users.rb
Comment (OsamaSayegh) [reply db/post_migrate/20210624023831_remove_highest_seen_post_number_from_topic_users.rb:5]: I have the same question/concern as Martin here, do we need to drop the column now? I see you've added the column to `ignored_columns` with a comment to drop the column in Jan next year, so I'm a little bit confused.

File: app/assets/javascripts/discourse/app/templates/components/topic-post-badges.hbs
Comment (OsamaSayegh) [inline app/assets/javascripts/discourse/app/templates/components/topic-post-badges.hbs:1]: I grepped for the name of this component in all-the-themes/plugins and I see it and `topic-post-badges.hbr` are referenced in a few themes/plugins. This PR renames a couple of attributes that this component receives so we probably should go through official themes/plugins and update them.

File: db/post_migrate/20210624023831_remove_highest_seen_post_number_from_topic_users.rb
Comment (SamSaffron) [reply db/post_migrate/20210624023831_remove_highest_seen_post_number_from_topic_users.rb:5]: Nah, we can easily get the data back if we need it, we can rebuild it from post_timings.

**Comment by SamSaffron** [comment]
very excited about this change but I share the concern that we are going to be on theme cleanup for a bit. 

I recommend merging this on Monday and doing all the prep work on existing themes. Can a cross compatible change be made to themes that rewrite hbs?

**Review by SamSaffron — APPROVED** [review]

File: spec/fixtures/json/sam-s-simple-theme.dcstyle.json
Comment (tgxworld) [reply spec/fixtures/json/sam-s-simple-theme.dcstyle.json:1]: A key in the serializer payload has changed xD

**Comment by tgxworld** [comment]
@SamSaffron @martin-brennan @jjaffeux @OsamaSayegh I've updated the PR in such a way that it'll be backwards compatible with themes. Can you guys have a look at this again?

**Comment by tgxworld** [comment]
Ping @SamSaffron @martin-brennan @OsamaSayegh I've addressed the initial concerns with plugins and themes by making the changes backwards compatible. Do you mind reviewing this today or tomorrow so that I can merge early in the week in case there are bugs I need to fix? Thank you!

**Comment by SamSaffron** [comment]
Had another looks, it looks good to me! Feel free to merge once you get another approval!

**Review by martin-brennan — APPROVED** [review]
@tgxworld LGTM :)

**Pull request merged**
